testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    cluster::Cluster,
+    instance::Instance,
+    instance::validator_index,
+    health::{HealthCheck, HealthCheckContext},
+};
+use async_trait::async_trait;
+use futures::future::join_all;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+use std::env;
+
+pub static THRESHOLD: Lazy<i64> = Lazy::new(||
+    if let Ok(v) = env::var("FULL_NODE_HEALTH_THRESHOLD") {
+        v.parse().expect("Failed to parse FULL_NODE_HEALTH_THRESHOLD")
+    } else {
+        20000_i64
+    });
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster) -> FullNodeHealthCheck {
+        Self { cluster }
+    }
+}
+
+async fn get_version(instance: &Instance) -> (String, i64) {
+    let res = instance
+        .debug_interface_client()
+        .get_node_metric("libra_state_sync_committed_version{}")
+        .await;
+    let content = match res {
+        Ok(res) => res.unwrap_or_default(),
+        _ => 0i64,
+    };
+    (instance.peer_name().clone(), content)
+}
+
+#[async_trait]
+impl HealthCheck for FullNodeHealthCheck {
+    async fn verify(&mut self, ctx: &mut HealthCheckContext) {
+        let validator = self.cluster.validator_instances();
+        let fullnode = self.cluster.fullnode_instances();
+
+        let futures = validator.iter().map(get_version);
+        let val_latest_version = join_all(futures).await;
+        let val_latest_version: HashMap<String, i64> = val_latest_version.into_iter().collect();
+
+        let futures = fullnode.iter().map(get_version);
+        let fullnode_latest_version = join_all(futures).await;
+        let fullnode_latest_version: HashMap<String, i64> =
+            fullnode_latest_version.into_iter().collect();
+
+        for (peer_name, fullnode_version) in fullnode_latest_version {
+            let index = validator_index(&peer_name);
+            let val_version = val_latest_version[&format!("val-{}", index)];
```

Comment:
After this you going to have `instance` instead of `peer_name` here, and can use `let val_version = val_latest_version[instance.validator_index()]`

Reply:
my actual problem is we may have multi-fullnode with same index map to same validator, we have to verify all fullnode status with that one validator, but return Map<Instance, i64> is a good way to solve it